File: solutions/go/tree-building/1/tree_building.go
package tree

import (
	"errors"
	"sort"
)

type Record struct {
	ID     int
	Parent int
}

type Node struct {
	ID       int
	Children []*Node
	Parent   int
}

// Build creates a tree given a list of records.
func Build(records []Record) (*Node, error) {
	if len(records) == 0 {
		return nil, nil
	}

	sort.Slice(records, func(i, j int) bool {
		return records[i].ID < records[j].ID
	})

	for i := 1; i < len(records); i++ {
		if records[i].ID != records[i-1].ID+1 {
			return nil, errors.New("non-continuous IDs detected")
		}
	}

	nodeById := make(map[int]*Node)

	for i := 0; i < len(records); i++ {
		if _, ok := nodeById[records[i].ID]; ok {
			return nil, errors.New("duplicate node")
		}
		nodeById[records[i].ID] = &Node{
			ID:     records[i].ID,
			Parent: records[i].Parent,
		}
	}

	rootNode, ok := nodeById[0]

	if !ok {
		return nil, errors.New("no root node given")
	} else if rootNode.Parent != 0 {
		return nil, errors.New("root node has parent")
	}

	for _, node := range nodeById {
		if node.ID == 0 {
			continue
		}

		if node.ID <= node.Parent {
			return nil, errors.New("invalid parent")
		}

		parentNode := nodeById[node.Parent]

		index := sort.Search(len(parentNode.Children), func(i int) bool {
			return parentNode.Children[i].ID > node.ID
		})

		parentNode.Children = append(parentNode.Children, nil)
		copy(parentNode.Children[index+1:], parentNode.Children[index:])
		parentNode.Children[index] = node
	}

	return rootNode, nil
}

File: solutions/python/tree-building/1/tree_building.py
class Record:
    def __init__(self, record_id, parent_id):
        self.record_id = record_id
        self.parent_id = parent_id


class Node:
    def __init__(self, node_id):
        self.node_id = node_id
        self.children = []


def BuildTree(records):
    items = {r.record_id: Node(r.record_id) for r in records}
    root = items.get(0, None)

    if any(n not in items for n in range(len(records))):
        raise ValueError("Record id is invalid or out of order.")

    for r in sorted(records, key=lambda r: r.record_id):
        if r.parent_id > r.record_id:
            raise ValueError("Node parent_id should be smaller than it's record_id.")

        if r.record_id != 0:
            if r.record_id == r.parent_id:
                raise ValueError("Only root should have equal record and parent id.")

            items[r.parent_id].children.append(items[r.record_id])

    return root
